community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/HeapSizeSanityChecker.java

```diff
@@ -0,0 +1,86 @@
+
+import org.neo4j.kernel.impl.store.format.RecordFormats;
+import org.neo4j.kernel.impl.util.OsBeanUtil;
+import org.neo4j.unsafe.impl.batchimport.cache.MemoryStatsVisitor;
+import org.neo4j.unsafe.impl.batchimport.input.Input;
+
+import static org.neo4j.kernel.impl.util.OsBeanUtil.VALUE_UNAVAILABLE;
+import static org.neo4j.unsafe.impl.batchimport.ImportMemoryCalculator.estimatedCacheSize;
+import static org.neo4j.unsafe.impl.batchimport.ImportMemoryCalculator.optimalMinimalHeapSize;
+
+/**
+ * Sanity checking of {@link org.neo4j.unsafe.impl.batchimport.input.Input.Estimates} against heap size and free memory.
+ * Registers warnings onto a {@link ImportLogic.Monitor}.
+ */
+public class HeapSizeSanityChecker
+{
+    private final ImportLogic.Monitor monitor;
+    private final LongSupplier freeMemoryLookup;
+    private final LongSupplier actualHeapSizeLookup;
+
+    HeapSizeSanityChecker( ImportLogic.Monitor monitor )
+    {
+        this( monitor, OsBeanUtil::getFreePhysicalMemory, Runtime.getRuntime()::maxMemory );
+    }
+
+    public HeapSizeSanityChecker( ImportLogic.Monitor monitor, LongSupplier freeMemoryLookup, LongSupplier actualHeapSizeLookup )
+    {
+        this.monitor = monitor;
+        this.freeMemoryLookup = freeMemoryLookup;
+        this.actualHeapSizeLookup = actualHeapSizeLookup;
+    }
+
+    public void sanityCheck( Input.Estimates inputEstimates, RecordFormats recordFormats, MemoryStatsVisitor.Visitable baseMemory,
+            MemoryStatsVisitor.Visitable... memoryVisitables )
+    {
+        // At this point in time the store hasn't started so it won't show up in free memory reported from OS,
+        // i.e. we have to include it here in the calculations.
+        long estimatedCacheSize = estimatedCacheSize( baseMemory, memoryVisitables );
+        long freeMemory = freeMemoryLookup.getAsLong();
+        long optimalMinimalHeapSize = optimalMinimalHeapSize( inputEstimates, recordFormats );
+        long actualHeapSize = actualHeapSizeLookup.getAsLong();
+        boolean freeMemoryIsKnown = freeMemory != VALUE_UNAVAILABLE;
+
+        // Check if there's enough memory for the import
+        if ( freeMemoryIsKnown && actualHeapSize + freeMemory < estimatedCacheSize + optimalMinimalHeapSize )
+        {
+            monitor.insufficientAvailableMemory( estimatedCacheSize, optimalMinimalHeapSize, freeMemory );
+            return; // there's likely not available memory, no need to warn about anything else
+        }
+
+        // Check if the heap is big enough to handle the import
+        if ( actualHeapSize < optimalMinimalHeapSize )
+        {
+            monitor.insufficientHeapSize( optimalMinimalHeapSize, actualHeapSize );
+            return; // user have been warned about heap size issue
+        }
+
+        // Check if heap size could be tweaked
+        if ( ((freeMemoryIsKnown && freeMemory < estimatedCacheSize) || !freeMemoryIsKnown) && actualHeapSize > optimalMinimalHeapSize * 1.2 )
```

Comment:
expression can be simplified to `( (!freeMemoryIsKnown || freeMemory < estimatedCacheSize) && (actualHeapSize > (optimalMinimalHeapSize * 1.2)) )`